src/metabase/util/string.clj

```diff
@@ -56,35 +57,38 @@
   [s] (try (java.util.UUID/fromString s) true
            (catch Exception _e false)))
 
-(defn elide
-  "Elides the string to the specified length, adding '...' if it exceeds that length."
-  ^String [^String s max-length]
-  (if (> (count s) max-length)
-    (str (subs s 0 (- max-length 3)) "...")
-    s))
-
-(defn- remove-chars
-  "Removes individual chars until it fits in the required bytes"
-  ^String [^String s max-bytes]
-  (if (nil? s)
-    s
-    (loop [index (count s)]
-      (let [^String truncated (subs s 0 index)
-            bytes (.getBytes truncated "UTF-8")]
-        (if (<= (count bytes) max-bytes)
-          truncated
-          (recur (dec index)))))))
+(defn- continuation-byte? [b]
+  (= (bit-and b 0xC0) 0x80))
 
 (defn limit-bytes
   "Limits the string to the given number of bytes, ensuring it's still a valid UTF-8 string"
   ^String [^String s max-bytes]
   (if (nil? s)
     s
-    (let [bytes (.getBytes s "UTF-8")]
-      (if (<= (count bytes) max-bytes)
-        s
-        ;; first do big first-pass at truncating, then truncate the rest of the way to preserve a valid string
-        (remove-chars (String. (byte-array (take max-bytes bytes)) "UTF-8") max-bytes)))))
+    (if (<= (Utf8/encodedLength s) max-bytes)
+      s
+      ;; first do big first-pass at truncating, then truncate the rest of the way to preserve a valid string
+      (let [bytes (.getBytes s "UTF-8")
+            end   (loop [pos max-bytes]
+                    (if-not (and (pos? pos)
+                                 (continuation-byte? (aget bytes pos)))
+                      pos
+                      (recur (dec pos))))]
```

Comment:
nit: having `(cond (zero? pos) 0 (continuation-byte? ...) pos :else (recur ...))` feels more readable to me.